### ISSUE TITLE
DROP PARTITION hangs on v19.1.14

### ISSUE THREAD

**Issue by kmatt** [ClickHouse/ClickHouse#4282]
**Describe the bug**

Dropping partition operations never complete, don't cancel with KILL QUERY, and block other read / write operations on the tables being altered.

Service cannot be stopped and shutdown requires `kill -9`

**How to reproduce**
* Which ClickHouse server version to use: 19.1.6-1 (54413) from Altinity RPM repo
* Which interface to use: `clickhouse-client`
* OS: CentOS 7.4.1708

* Queries to run that lead to unexpected result:

1) Populate table with Kafka engine through materialized view.
2) Select oldest partition from system.parts `WHERE active`:
3) `ALTER TABLE table DROP PARTITION ('2018-09-23','id1');`

**Expected behavior**
With v18.16.1 the partitions dropped with a little extra system load, in 20-30 seconds.

With v19.1.6-1, inserts to table being altered stop, ALTER TABLE cannot be killed, `clickhouse-server` does not respond to service shutdown. ALTER query shows a running state after 30 minutes.

Reverting to v18.16.1 on same cluster allows partition drops to complete in normal time.

**Error message and/or stacktrace**
No exceptions appear logs related to drop partition. Merges appear to continue in background, but inserts do not.

**Comment by kmatt** [comment]
Example table structure for shards and Kafka consumers. Distributed engine is not used in this case:

```sql
CREATE TABLE logs_shard
(
    shard_key UInt64,
    id String,
    dateTime DateTime,
    ...
)
ENGINE = MergeTree
PARTITION BY (toStartOfHour(dateTime), id)
ORDER BY (cid, dateTime)
SETTINGS index_granularity = 8192;

CREATE TABLE logs_kafka
(
    id String,
    cid String,
    date String,
    time String,
    ...
)
ENGINE = Kafka
SETTINGS kafka_broker_list = '{3 brokers}', kafka_topic_list = 'logs', kafka_group_name = 'clickhouse0', kafka_format = 'JSONEachRow', kafka_num_consumers = 3;


CREATE MATERIALIZED VIEW logs_consumer TO logs_shard
(
    shard_key UInt64,
    id String,
    cid String,
    dateTime DateTime,
    ...
) AS
SELECT
    reinterpretAsUInt64(cid) AS shard_key,
    id,
    cid,
    toDateTime(concat(date, ' ', time)) AS dateTime,
    ...
FROM logs_kafka;
```

**Comment by alexey-milovidov** [comment]
Does the issue anyhow related to the Kafka table?
(Does it reproduce if you first remove the Kafka table and materialized view and only then issue an ALTER command?)

Could you please install `clickhouse-common-dbg` package, reproduce the issue and run
`sudo gdb -batch -ex 't apply all bt' -p $(pidof clickhouse-server) 2>&1 | tee trace`
and attach the corresponding `trace` file?

**Comment by kmatt** [comment]
Would `clickhouse-common-dbg` be the same as `clickhouse-debuginfo`, noting the versions are not equivalent?

```
Available Packages
clickhouse-client.x86_64                                 19.1.6-1.el7                              Altinity_clickhouse
clickhouse-common-static.x86_64                          19.1.6-1.el7                              Altinity_clickhouse
clickhouse-compressor.x86_64                             1.1.54336-3.el7                           Altinity_clickhouse
clickhouse-debuginfo.x86_64                              18.14.15-1.el7                            Altinity_clickhouse
clickhouse-mysql.noarch                                  0.0.20180319-1                            Altinity_clickhouse
clickhouse-odbc.x86_64                                   20180903-1                                Altinity_clickhouse
clickhouse-server.x86_64                                 19.1.6-1.el7                              Altinity_clickhouse
clickhouse-server-common.x86_64                          19.1.6-1.el7                              Altinity_clickhouse
clickhouse-test.x86_64                                   19.1.6-1.el7                              Altinity_clickhouse
```

**Comment by kmatt** [comment]
In preparation for the test, I experienced a similar hang of a DROP TABLE of a Kafka engine type on v18.16.1. Attaching the trace for this condition here, captured at 10 minutes into the drop process.

[trace-18.16.1-github4282.txt](https://github.com/yandex/ClickHouse/files/2870053/trace-18.16.1-github4282.txt)

**Comment by kmatt** [comment]
Initial testing with v19.3.3 (54415) is allowing partition drops to occur as they did with v18.

**Comment by kmatt** [comment]
re https://github.com/yandex/ClickHouse/issues/4282#issuecomment-464122556 

Dropping of Kafka engine tables is also working as normal on v19.3.3

**Comment by kmatt** [comment]
Seeing this behavior again on 19.1.4 and 19.5.3.8

**Comment by kmatt** [comment]
> Is the issue anyhow related to the Kafka table?
> (Does it reproduce if you first remove the Kafka table and materialized view and only then issue an ALTER command?)
> 
> Could you please install `clickhouse-common-dbg` package, reproduce the issue and run
> `sudo gdb -batch -ex 't apply all bt' -p $(pidof clickhouse-server) 2>&1 | tee trace`
> and attach the corresponding `trace` file?

@alexey-milovidov Current situation I'm seeing in v19.1.14 in Docker is when Kafka engine tables and their materialized views are dropped, the clickhouse-server process crashes, and the only way to restart is to delete the metadata files from the data path. I can't find anything relevant in the logs even at debug logging level.

Attaching gdb does not seem like an option here. Any other diagnostic suggestions?

**Comment by filimonov** [comment]
I believe it was caused by Kafka issue 
```
Thread 30 (Thread 0x7f62eccff700 (LWP 851)):
#0  0x00007f690b824f57 in pthread_join () from /lib64/libpthread.so.0
#1  0x000000000651c005 in thrd_join ()
#2  0x000000000649585f in rd_kafka_destroy ()
#3  0x0000000005005182 in DB::StorageKafka::shutdown() ()
#4  0x0000000004db65dd in DB::InterpreterDropQuery::executeToTable(std::string&, std::string&, DB::ASTDropQuery::Kind, bool, bool) ()
#5  0x0000000004db8e6e in DB::InterpreterDropQuery::execute() ()
#6  0x0000000004eadddc in DB::executeQueryImpl(char const*, char const*, DB::Context&, bool, DB::QueryProcessingStage::Enum) ()
#7  0x0000000004eafec0 in DB::executeQuery(std::string const&, DB::Context&, bool, DB::QueryProcessingStage::Enum) ()
#8  0x0000000002bdc277 in DB::TCPHandler::runImpl() ()
#9  0x0000000002bdd4dc in DB::TCPHandler::run() ()
#10 0x0000000005fc3cef in Poco::Net::TCPServerConnection::start() ()
#11 0x0000000005fc40b6 in Poco::Net::TCPServerDispatcher::run() ()
#12 0x00000000062755a7 in Poco::PooledThread::run() ()
#13 0x0000000006271768 in Poco::ThreadImpl::runnableEntry(void*) ()
#14 0x0000000006a7fa8f in execute_native_thread_routine ()
#15 0x00007f690b823e25 in start_thread () from /lib64/libpthread.so.0
#16 0x00007f690b04643d in clone () from /lib64/libc.so.6
```

It was fixed in version >20.3.11 / >20.5. 

See
https://github.com/ClickHouse/ClickHouse/issues/7260
https://github.com/ClickHouse/ClickHouse/issues/11944

**Comment by cangyin** [comment]
> I believe it was caused by Kafka issue
> 
> ```
> Thread 30 (Thread 0x7f62eccff700 (LWP 851)):
> #0  0x00007f690b824f57 in pthread_join () from /lib64/libpthread.so.0
> #1  0x000000000651c005 in thrd_join ()
> #2  0x000000000649585f in rd_kafka_destroy ()
> #3  0x0000000005005182 in DB::StorageKafka::shutdown() ()
> #4  0x0000000004db65dd in DB::InterpreterDropQuery::executeToTable(std::string&, std::string&, DB::ASTDropQuery::Kind, bool, bool) ()
> #5  0x0000000004db8e6e in DB::InterpreterDropQuery::execute() ()
> #6  0x0000000004eadddc in DB::executeQueryImpl(char const*, char const*, DB::Context&, bool, DB::QueryProcessingStage::Enum) ()
> #7  0x0000000004eafec0 in DB::executeQuery(std::string const&, DB::Context&, bool, DB::QueryProcessingStage::Enum) ()
> #8  0x0000000002bdc277 in DB::TCPHandler::runImpl() ()
> #9  0x0000000002bdd4dc in DB::TCPHandler::run() ()
> #10 0x0000000005fc3cef in Poco::Net::TCPServerConnection::start() ()
> #11 0x0000000005fc40b6 in Poco::Net::TCPServerDispatcher::run() ()
> #12 0x00000000062755a7 in Poco::PooledThread::run() ()
> #13 0x0000000006271768 in Poco::ThreadImpl::runnableEntry(void*) ()
> #14 0x0000000006a7fa8f in execute_native_thread_routine ()
> #15 0x00007f690b823e25 in start_thread () from /lib64/libpthread.so.0
> #16 0x00007f690b04643d in clone () from /lib64/libc.so.6
> ```
> 
> It was fixed in version >20.3.11 / >20.5.
> 
> See #7260 #11944

Just out of curiosity, I am getting almost the same ouput from gdb. But why gdb is not resolving line numbers with  clickhouse-common-dbg package installed ?
